internal/configfsm/fsm.go

```diff
@@ -0,0 +1,278 @@
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case Missing:
+		err = fsm.Fire(MissingToMissing, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case ClusterNotReady:
+		err = fsm.Fire(ClusterNotReadyToClusterNotReady, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case InProgress:
+		err = fsm.Fire(InProgressToInProgress, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case TimedOut:
+		err = fsm.Fire(TimedOutToTimedOut, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case Completed:
+		err = fsm.Fire(CompletedToCompleted, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case OutOfDate:
+		err = fsm.Fire(OutOfDateToOutOfDate, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	}
+	// Update the current state
+	aState, err = fsm.State(ctx)
+	if err != nil {
+		return aState, fmt.Errorf("could not get state, err: %w", err)
+	}
+	return aState, nil
+}
+
+// InitFSM Initialize the state machine algorithm
+func InitFSM(state string) (fsm *stateless.StateMachine, err error) {
+	fsm = stateless.NewStateMachine(state)
+	fsm.Configure(Start).Permit(StartToMissing, Missing)
+	fsm.Configure(Missing).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering " + Missing)
+			fsmHelper.ResetNonCompliantAt()
+			if fsmHelper.ArePoliciesMatched() {
+				return fsm.Fire(MissingToClusterNotReady, fsmHelper)
+			}
+			return nil
+		}).
+		Permit(MissingToClusterNotReady, ClusterNotReady).
+		PermitReentry(MissingToMissing)
+	fsm.Configure(ClusterNotReady).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering " + ClusterNotReady)
+			if !fsmHelper.ArePoliciesMatched() {
+				return fsm.Fire(ClusterNotReadyToMissing, fsmHelper)
+			}
+			if fsmHelper.IsClusterReady() ||
+				fsmHelper.IsTimedOut() {
+				return fsm.Fire(ClusterNotReadyToInProgress, fsmHelper)
+			}
+			return nil
+		}).
+		Permit(ClusterNotReadyToMissing, Missing).
+		Permit(ClusterNotReadyToInProgress, InProgress).
+		PermitReentry(ClusterNotReadyToClusterNotReady)
+	fsm.Configure(InProgress).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering " + InProgress)
+			if fsmHelper.IsNonCompliantAtZero() ||
+				fsmHelper.IsAllPoliciesCompliant() ||
+				(!fsmHelper.IsNonCompliantPolicyInEnforce() && !fsmHelper.IsAllPoliciesCompliant()) {
+				fsmHelper.SetNonCompliantAtNow()
+			}
+			if !fsmHelper.ArePoliciesMatched() {
```

Comment:
It looks like we can group this one with `|| !fsmHelper.IsClusterReady()` since they both transition to `ClusterNotReady`

Reply:
we can't group these 2 because the "fsmHelper.IsTimedOut()" must be evaluated first. Basically if we are in timeout we should stay in TimedOut even if the cluster is not ready